commit ecaf18c15aac8bb9bed7b7aa0e382fe252e275d5
Author: Eric Paris <eparis@redhat.com>
Date:   Tue Dec 4 23:45:31 2007 -0800

    VM/Security: add security hook to do_brk
    
    Given a specifically crafted binary do_brk() can be used to get low pages
    available in userspace virtual memory and can thus be used to circumvent
    the mmap_min_addr low memory protection.  Add security checks in do_brk().
    
    Signed-off-by: Eric Paris <eparis@redhat.com>
    Acked-by: Alan Cox <alan@redhat.com>
    Cc: Stephen Smalley <sds@tycho.nsa.gov>
    Cc: James Morris <jmorris@namei.org>
    Cc: Chris Wright <chrisw@sous-sol.org>
    Signed-off-by: Andrew Morton <akpm@linux-foundation.org>
    Signed-off-by: Linus Torvalds <torvalds@linux-foundation.org>

Adjusted to apply to Debian's 2.6.18 by dann frazier <dannf@debian.org>

diff -urpN linux-source-2.6.18.orig/mm/mmap.c linux-source-2.6.18/mm/mmap.c
--- linux-source-2.6.18.orig/mm/mmap.c	2008-01-15 16:46:27.000000000 -0700
+++ linux-source-2.6.18/mm/mmap.c	2008-01-16 00:28:42.000000000 -0700
@@ -1883,6 +1883,10 @@ unsigned long do_brk(unsigned long addr,
 	if ((addr + len) > TASK_SIZE || (addr + len) < addr)
 		return -EINVAL;
 
+	error = security_file_mmap(0, 0, 0, 0, addr, 1);
+	if (error)
+		return error;
+
 	flags = VM_DATA_DEFAULT_FLAGS | VM_ACCOUNT | mm->def_flags;
 
 	error = arch_mmap_check(addr, len, flags);
